### DISCUSSION TITLE
Handling collections

### DISCUSSION THREAD

**Discussion by slimshader** [codewriter-packages/UniMob#5] [Q&A] [answered]
Hi,

first of all: fantastic project, always great to see solutions that promote declarative designs. Looking forward to updates to UniMob.UI too.

Question tho: do I understand correctly that collections (like Todo[] array) have to be replaced entirely for Atom to detect changes? (So basically it should be be treated as immutable collection) As in this would not work as expected, correct?:

```
[Atom] public List<Todo> Todos { get; set; } = new List<Todo>();
incrementButton.onClick.AddListener(() =>
{
       Todos.Add(new Todo());
});

Atom.Reaction(() => counterText.text = "Unifinished: " + UnfinishedTodoCount);
```

**Comment by vanifatovvlad** [comment, marked as answer]
Hi,

Yes that's right. Atom doesn't track object modification. Reaction will be triggered only when new value has been set for atom.

However, you can write your own collection type that will trigger reactions when new items are added.

```csharp
[Atom] public AtomList<Todo> Todos { get; set; } = new AtomList<Todo>();
incrementButton.onClick.AddListener(() =>
{
       Todos.Add(new Todo());
});

Atom.Reaction(() => counterText.text = "Unifinished: " + UnfinishedTodoCount);
```
```csharp
public class AtomList<T> : IEnumerable<T>
{
    private readonly MutableAtom<List<T>> _list = Atom.Value(new List<T>());

    public void Add(T value)
    {
        _list.Value.Add(value);
        _list.Invalidate();
    }

    public IEnumerator<T> GetEnumerator() => _list.Value.GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}
```

Here `_list.Invalidate()` method will trigger reactions in the same way as if the collection were replaced with a new one.

**Comment by slimshader** [comment]
Thanks, I see, Invalidate() is the key to custom atom semantics.

**Comment by slimshader** [comment]
Hey, sorry for reviving quite old thread but Invalidate() does nto seem to trigger Reactions for me, my AtomList implementation:

[gist](https://gist.github.com/slimshader/d35a20042f846e51174acff3bdc382ec)

calling Add() or Remove() does not trigger the Reacton

**Reply by vanifatovvlad** [reply]
I looked at the code and it works. Perhaps the problem is in the incorrect usage and not in the implementation of the AtomList?
